### PR TITLE
fix(reset): Improve cycle detector

### PR DESCRIPTION
**Pull request by idsulik**
Improved the cycle detector and added one more test case.  

https://github.com/docker/compose/issues/12247#issuecomment-2459965906

### CODE REVIEW

File: loader/reset.go [hunk 1]
```diff
@@ -144,20 +145,46 @@ func (p *ResetProcessor) applyNullOverrides(target any, path tree.Path) error {
...
+	}
+
+	p.visitedNodes[node] = append(paths, pathStr)
 	return nil
 }
+
+// areInDifferentServices checks if two paths are in different service definitions
+func areInDifferentServices(path1, path2 string) bool {
+	// Split paths into components
+	parts1 := strings.Split(path1, ".")
```
Comment (ndeloof) [inline loader/reset.go:178]: nit: could use path.Parts() if you don't convert it to a string

Comment (idsulik) [reply loader/reset.go:178]: the function accepts strings and the field also has a list of strings `visitedNodes map[*yaml.Node][]string`, do you suggest to use path everywhere? or to convert the strings inside the func into paths and use `Parts()`? it looks like overhead, isn't it?

Comment (ndeloof) [reply loader/reset.go:178]: some overhead indeed, but my point is that `Path` abstraction allows to ignore "implementation details" like the dot notation and `[]` or `*` patterns. I had in mind at some point we might want to use an alternate implementation vs `string` as we actually have to split this string many times during parsing.
From an encapsulation point of view, I'd prefer we add more utility func to `Path` to support required submatching. 
ANYWAY I'm fine we merge this PR as a quick fix for this issue, and revisit this later

**Review by ndeloof — APPROVED** [review]

**Review by glours — APPROVED** [review]
Let's do a follow up PR for the improvements

File: loader/reset.go (hunk 1 above)
Comment (idsulik) [reply loader/reset.go:178]: @ndeloof agree, that sounds reasonable.

**Pull request merged**